presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedHeap.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.type.Type;
+
+public class TypedHeap
+{
+    private static final int COMPACT_THRESHOLD_BYTES = 32768;
+    private static final int COMPACT_THRESHOLD_RATIO = 3; // when 2/3 of elements in heapBlockBuilder is unreferenced, do compact
+
+    private final BlockComparator comparator;
+    private final Type type;
+    private final int capacity;
+
+    private int size;
+    private final int[] heapIndex;
+    private BlockBuilder heapBlockBuilder;
+
+    public TypedHeap(BlockComparator comparator, Type type, int capacity)
+    {
+        this.comparator = comparator;
+        this.type = type;
+        this.capacity = capacity;
+        this.heapIndex = new int[capacity];
+        this.heapBlockBuilder = type.createBlockBuilder(new BlockBuilderStatus(), capacity);
+    }
+
+    public int getCapacity()
+    {
+        return capacity;
+    }
+
+    public long getEstimatedSize()
+    {
+        return heapBlockBuilder.getRetainedSizeInBytes() + capacity * Integer.BYTES;
+    }
+
+    public boolean isEmpty()
+    {
+        return size == 0;
+    }
+
+    public void writeAll(BlockBuilder resultBlockBuilder)
+    {
+        for (int i = 0; i < size; i++) {
+            heapBlockBuilder.writePositionTo(heapIndex[i], resultBlockBuilder);
```

Comment:
usually I see `Type.appendTo` used instead of `writePositionTo` followed by `closeEntry`, which is a little safer because it handles null positions. Perhaps we should think about removing `Type.appendTo` and providing a utility method that does the check?